### PR TITLE
Edit Python Port Properties

### PR DESCRIPTION
**Pull request by dnenov**
### Purpose

The goal of this PR is to allow users to **rename** Python imports and outports in order to create more user-friendly and descriptive content for their custom code. 

This [Figma ](https://www.figma.com/file/1q7EWQGYO7pPDhyLf8nuwW/Python-node-editor-restyling?node-id=393%3A63878&t=qTOpC79Sldt14gag-0)link is created for visual guidance. 

#### Current WIP
image

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Release Notes

WIP

- added a new button to the OutPortContextMenu to allow users to Edit Port Properties
- for the time being, it is only enabled and visible in case of Python Node Model (in future this check can be removed allowing other nodes to use this functionality as well)

### Reviewers

@mjkkirschner 
@Amoursol 
@sm6srw 

### FYIs

@Jingyi-Wen

### CODE REVIEW

File: src/DynamoCoreWpf/UI/Converters.cs [hunk 1]
```diff
@@ -3544,6 +3544,28 @@ public object ConvertBack(object value, Type targetType, object parameter, Cultu
         }
     }
 
+    [ValueConversion(typeof(PortType), typeof(string))]
```
Comment (reddyashish) [inline src/DynamoCoreWpf/UI/Converters.cs:3547]: Add some comments to a public converter.

File: src/DynamoCoreWpf/UI/Prompts/PortPropertiesEditPrompt.xaml [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+                         Style="{DynamicResource HintingInputStyle}"
+                         TextWrapping="Wrap"
+                         VerticalScrollBarVisibility="Auto" />
+            </StackPanel>
+        </Border>
+
+        <!--  Lower Section: Buttons  -->
+        <DockPanel Grid.Row="2" Margin="0,0,25,25">
+            <Button Name="cancelButton"
+                    MinWidth="85"
```
Comment (reddyashish) [inline src/DynamoCoreWpf/UI/Prompts/PortPropertiesEditPrompt.xaml:252]: According to the figma, the cancel button is supposed to be on the right. Can you confirm that?

File: src/DynamoCoreWpf/UI/Prompts/PortPropertiesEditPrompt.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+            {
+                var f = WpfUtilities.FindUpVisualTree<DynamoView>(this);
+                if (f != null) return f;
+
+                return null;
+            }
+        }
+
+
+        private string errorString = string.Empty;
```
Comment (reddyashish) [inline src/DynamoCoreWpf/UI/Prompts/PortPropertiesEditPrompt.xaml.cs:85]: Move the private properties to the top of class for better readability.

File: src/DynamoCoreWpf/ViewModels/Core/InPortViewModel.cs [hunk 4]
```diff
@@ -284,6 +305,54 @@ public DelegateCommand MouseLeftButtonDownOnLevelCommand
...
+        }
+        /// <summary>
+        /// Used by the 'Edit Port Properties' button in the node output context menu.
+        /// Triggers the Port Properties Panel
+        /// </summary>
+        private void EditPortProperties(object parameter)
+        {
+            var wsViewModel = node.WorkspaceViewModel;
+
+            //wsViewModel.CancelActiveState();
```
Comment (reddyashish) [inline src/DynamoCoreWpf/ViewModels/Core/InPortViewModel.cs:328]: Remove commented code.

File: src/DynamoCoreWpf/ViewModels/Core/OutPortViewModel.cs [hunk 5]
```diff
@@ -296,6 +340,42 @@ private void HideConnections(object parameter)
...
 
+        /// <summary>
+        /// Used by the 'Edit Port Properties' button in the node output context menu.
+        /// Triggers the Port Properties Panel
+        /// </summary>
+        private void EditPortProperties(object parameter)
+        {
+            var wsViewModel = node.WorkspaceViewModel;
+
+            //wsViewModel.CancelActiveState();
```
Comment (reddyashish) [inline src/DynamoCoreWpf/ViewModels/Core/OutPortViewModel.cs:351]: Remove commented code here.

**Review by reddyashish — COMMENTED** [review]
Conflicts need to be resolved. LGTM with some comments.

File: src/DynamoCoreWpf/UI/Converters.cs (hunk 1 above)
Comment (dnenov) [reply src/DynamoCoreWpf/UI/Converters.cs:3550]: Yup!

File: src/DynamoCoreWpf/UI/Prompts/PortPropertiesEditPrompt.xaml (hunk 2 above)
Comment (dnenov) [reply src/DynamoCoreWpf/UI/Prompts/PortPropertiesEditPrompt.xaml:252]: image

Comment (reddyashish) [reply src/DynamoCoreWpf/UI/Prompts/PortPropertiesEditPrompt.xaml:252]: Thanks @dnenov

File: src/DynamoCoreWpf/UI/Prompts/PortPropertiesEditPrompt.xaml.cs [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+            }
+
+            return false;
+        }
+
+
+        public bool IsPortNameInValid(string portName)
+        {
+            // Some other extra characters that are to be checked. 
+            char[] invalidCharactersFileName = { '#', '%', '&', '.', ' ', '[', ']' };
```
[the remark this answers is not in the document]
Comment (dnenov) [reply src/DynamoCoreWpf/UI/Prompts/PortPropertiesEditPrompt.xaml.cs]: Done!

File: src/DynamoCoreWpf/ViewModels/Core/InPortViewModel.cs (hunk 4 above)
Comment (dnenov) [reply src/DynamoCoreWpf/ViewModels/Core/InPortViewModel.cs]: Done

**Comment by reddyashish** [comment]
passed: https://master-15.jenkins.autodesk.com/view/DYN/job/DYN-DevCI_Self_Service/1130/

**Review by reddyashish — APPROVED** [review]

**Pull request merged**